### DISCUSSION TITLE
React Props and State - Not really understanding this component

### DISCUSSION THREAD

**Discussion by BrysonMG** [nss-day-cohort-47/C47-Discussions#74] [Help Me] [answered]
We have just started learning React, and I'm slowly coming to understanding it, but one component we went over doesn't really make enough sense to me to be able to write my own without copying the example line for line. Sorry, the insert code feature doesn't want to put the whole thing inside the box.

---

`export const PropsAndState = ({ yourName }) => {

    let [countClicks, setCountClicks] = useState(0)
    const handleClick = () => {
        //make a copy of state, modify it, and then setState to the copy
        const newCountClicks = ++countClicks
        setCountClicks(newCountClicks)
    }
    return (
        <>
            <h3>Welcome, {yourName} </h3>
            <p>{countClicks}</p>
            <button onClick={(handleClick)}>Click Me</button>
        </>
    )
}`

---

So the questions that stand out to me right now are:

- The line that say `let [countClicks, setCountClicks] = useState(0)`
- - Where are countClicks and setCountClicks coming from?
- - Could we have called them anything?
- - Can I get an explanation of the syntax of the entire line perhaps? I don't really understand any part of this line
- - What is the 0 inside of useState there for?

**Comment by BrysonMG** [comment]
I got some help from Monica and understand that line a bit better now. I'd still be grateful to hear another person's explanation though. If nobody has commented by tomorrow, I'll try to hop on and write up an explanation based on what Monica told me for anyone else who is confused by this.

**Comment by BryanNilsen** [comment, marked as answer]
The line that says `let [countClicks, setCountClicks] = useState(0)`

**Where are countClicks and setCountClicks coming from?**
- `countClicks` and `setCountClicks` are the return values of calling `useState()` - this is a **React Hook**

**Could we have called them anything?**
- YES!
- `countClicks` is the variable name that YOU CHOOSE which will store some data into memory somewhere in the depths of the React codebase
- `setCountClicks` is the variable name that YOU CHOOSE to give to the **method** that allows you to update the **state** of the `countClicks` variable
- it is convention to just add `set` to the name of the state variable you just declared

**What is the 0 inside of useState there for?**
- This allows you to set the initial value for the `countClicks` variable
- In this case, we're setting it to zero by passing it into the `useState()` hook like this >> `useState(0)`

**Can I get an explanation of the syntax of the entire line perhaps? I don't really understand any part of this line**
- Basically, we need a way to store some data, set an initial value, and tell React to watch for changes to that data so it can re-render the component when that data changes.
- In this case, we want to store how many times a user clicks on a button and possibly display number back to the user.
- React gives us the `useState` hook to help us do that because their devs are _wicked smaaht_!
- So we call `useState()` and pass it an argument of '0' because that is our starting value.
- `useState` returns an array consisting of the value we passed in and the method to update that value. And so we can **destructure** that array on the left side of the `=` and declare variable names for each.
- so the first time we access the `countClicks` variable, it will have an initial value of zero
- if we want to update the value of`countClicks`, we cant do it directly like `countClicks = 1` because that is not how the hook works
- YOU HAVE TO USE THE `setCountClicks` method to update the state of `countClicks`
- something like `setCountClicks(1)` is totally cool and when that code runs, the value of `countClicks` gets updated from 0 to 1, triggering a re-render of the component
- Why does it work that way? [that's just the way it is...](https://www.youtube.com/watch?v=cOeKidp-iWo)

NOW, hopefully, when you read that line again, it will make a bit more sense! Took me quite a bit to wrap my head around it so don't worry. We're gonna use it a lot!!
